connectors/byoc.py

```diff
@@ -178,96 +178,142 @@ async def all_connectors(self):
             yield connector
 
 
-class SyncJob:
-    def __init__(self, elastic_index, connector_id, doc_source=None):
-        self.connector_id = connector_id
-        self.elastic_index = elastic_index
-        self.created_at = datetime.now(timezone.utc)
-        self.completed_at = None
-        self.filtering = Filter()
-        self.client = elastic_index.client
+class ESDocument:
+    def __init__(self, elastic_index, doc_source=None):
+        self.index = elastic_index
         if doc_source is None:
             doc_source = dict()
+        self.id = doc_source.get("_id")
+        self._source = doc_source.get("_source", {})
 
-        self.doc_source = doc_source
-        self.job_id = self.doc_source.get("_id")
-        self.status = str2e(self.doc_source.get("_source", {}).get("status"), JobStatus)
+    def get(self, *keys, default=None):
+        value = self._source
+        for key in keys:
+            if not isinstance(value, dict):
+                return default
+            value = value.get(key)
+        if value is None:
+            return default
+        return value
 
+    async def reload(self):
+        return await self.index.fetch_by_id(self.id)
```

Comment:
what happens if `self.id` is `None` (it's possible with the current constructor)

scenario:

```python
doc = ESDocument('index')
await doc.reload()  
```

we might want to add a safeguard

Reply:
What do you think we add to `__init__` of `ESDocument`?

```python
class InvalidDocumentSourceError(Exception):
    pass


class ESDocument:
    def __init__(self, elastic_index, doc_source=None):
        self.index = elastic_index
        if not isinstance(doc_source, dict):
            raise InvalidDocumentSourceError(f"Invalid doc_source type, expected: {dict.__name__}, got: {doc_source.__class__.__name__}")
        self.id = doc_source.get("_id")
        if not isinstance(self.id, str):
            raise InvalidDocumentSourceError(
                f"Invalid id type, expected: {str.__name__}, got: {self.id.__class__.__name__}")
        self._source = doc_source.get("_source", {})
        if not isinstance(self._source, dict):
            raise InvalidDocumentSourceError(
                f"Invalid source type, expected: {dict.__name__}, got: {self._source.__class__.__name__}")
```

Reply:
Eventually I didn't use `isinstance(doc_source, dict)` as the response from the ES call is an `ObjectApiResponse` object, not a `dict`. Here is the change: https://github.com/elastic/connectors-python/pull/460/commits/5d03016809dcc71b19b10fb38df21dbfd79efa46